pkg/tests/end_to_end_tests/trace_tree_test.go

```diff
@@ -0,0 +1,392 @@
+	exec, err = db.Exec(ctx, insertDecoy)
+	require.NoError(t, err, "Failed to insert decoy span: %v", err)
+	require.Equal(t, int64(1), exec.RowsAffected(), "Expected to insert 1 spans. Got: %v", exec.RowsAffected())
+}
+
+func testTraceTree(t *testing.T, ctx context.Context, db *pgxpool.Pool) {
+	const qry = `
+	select span_id, coalesce(parent_span_id, -1), lvl, path::text
+	from trace_tree('3dadb2bf-0035-433e-b74b-9075cc9260e8') 
+	order by span_id`
+
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Level        int64
+		Path         string
+	}
+
+	t.Run("trace_tree", func(t *testing.T) {
+		expectedResults := []result{
+			{1, -1, 1, "{1}"},
+			{2, 1, 2, "{1,2}"},
+			{3, 1, 2, "{1,3}"},
+			{4, 1, 2, "{1,4}"},
+			{5, 2, 3, "{1,2,5}"},
+			{6, 2, 3, "{1,2,6}"},
+			{7, 2, 3, "{1,2,7}"},
+			{8, 3, 3, "{1,3,8}"},
+			{9, 3, 3, "{1,3,9}"},
+			{10, 3, 3, "{1,3,10}"},
+			{11, 4, 3, "{1,4,11}"},
+			{12, 4, 3, "{1,4,12}"},
+			{13, 6, 4, "{1,2,6,13}"},
+			{14, 8, 4, "{1,3,8,14}"},
+			{15, 12, 4, "{1,4,12,15}"},
+			{16, 13, 5, "{1,2,6,13,16}"},
+			{17, 13, 5, "{1,2,6,13,17}"},
+			{18, 15, 5, "{1,4,12,15,18}"},
+		}
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running trace_tree: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			if !rows.Next() {
+				t.Fatalf("Expected a row for span id %v but got none", expected.SpanId)
+			}
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Level, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	})
+}
+
+func testUpstreamSpans(t *testing.T, ctx context.Context, db *pgxpool.Pool) {
+	type result struct {
+		SpanId       int64
+		ParentSpanId int64
+		Dist         int64
+		Index        int64
+		Path         string
+	}
+
+	eval := func(t *testing.T, qry string, expectedResults []result) {
+		rows, err := db.Query(ctx, qry)
+		require.NoError(t, err, "Error running upstream_spans: %v", err)
+		defer rows.Close()
+
+		for _, expected := range expectedResults {
+			if !rows.Next() {
+				t.Fatalf("Expected a row for span id %v but got none", expected.SpanId)
+			}
+
+			var actual = result{}
+			err = rows.Scan(&actual.SpanId, &actual.ParentSpanId, &actual.Dist, &actual.Index, &actual.Path)
+			require.NoError(t, err, "Failed to scan row. Expected row for span id %v: %v", expected.SpanId, err)
+
+			require.Equal(t, expected, actual)
+		}
+		require.False(t, rows.Next(), "Found more result rows than expected")
+	}
+
+	t.Run("upstream_spans", func(t *testing.T) {
+		const qry = `
+		select span_id, coalesce(parent_span_id, -1), dist, idx, path::text
+		from upstream_spans('3dadb2bf-0035-433e-b74b-9075cc9260e8', 13)
+		order by span_id`
+		expectedResults := []result{
+			{1, -1, 3, 4, "{13,6,2,1}"},
```

Comment:
hmm I think "{13,6,2,1}" should be "{1,2,6,13}" shouldn't we keep the order in the paths consistently upstream->downstream for all these functions?

Reply:
related question about whether dist should be positive or negative and relation with span_tree

Reply:
I'm not adamant that it stay the way I coded it, but let me explain my reasoning.
All the results are from the perspective of the span id passed as an argument. Paths going upstream (towards the root) have a negative distance, and paths going downstream have a positive distance. This makes it easy to determine the direction, and an `abs(dist)` is easy if you don't care. All the paths start with the span id that was passed with an arg. The index is the index of the span of the result row in the path array.

Reply:
Okay. In all of these functions, I rendered the path from the perspective of the span_id that was passed as an argument. It's the path, from the start span to the span identified by the result row.

It doesn't "have" to be that way, but that was my thinking.

Reply:
In `span_tree`, I made upstream distances negative and downstream distances positive so that you can easily determine the direction and an `abs(dist)` is easy if you need it.

We can ditch that idea and always use positive distances, or we can make the distances returned from `upstream_spans` negative if you want consistency.

Reply:
ahh I think the stuff with idx and dist makes sense now. I also see what you mean by "from the perspective of the span" but I do think having the paths go in a consistent order would make things more straightforward. Also in terms of documentation than a path is a defined thing with a defined order. So I think it would be better to change

Reply:
Okay, I don't feel strongly either way. Will make the change and update the tests

Reply:
what about `dist`? do you prefer it to be always positive, or what?

Reply:
`dist` changed to be negative for all upstream paths across the board for consistency